src/btree/bt_walk.c

```diff
@@ -624,3 +625,241 @@ __wti_tree_walk_skip(WT_SESSION_IMPL *session, WT_REF **refp, uint64_t *skipleaf
+
+    if (path_str) {
+        path_str[*path_str_offsetp] = 0;
+    }
+
+    return (WT_CLAMP(npos, 0.0, 1.0));
+}
+
+/*
+ * !!!
+ * __find_closest_page --
+ *     Find the closest suitable page according to flags.
+ *     - It should not be deleted.
+ *     - If WT_READ_CACHE is set, the page should be in memory.
+ */
+static int
+__find_closest_page(WT_SESSION_IMPL *session, WT_REF **refp, uint32_t flags)
+{
+    if (*refp == NULL || F_ISSET(*refp, WT_REF_FLAG_LEAF))
+        return (0);
+    LF_SET(WT_READ_SKIP_INTL);
+    return (__wt_tree_walk(session, refp, flags));
+}
+
+/*
+ * __page_from_npos_internal --
+ *     Go to a leaf page given its normalized position. Note that this function can return a "bad"
+ *     page (deleted, locked, etc). The caller of this function should walk the tree to find a
+ *     suitable page.
+ *
+ * NOTE: Must be called within WT_WITH_PAGE_INDEX or WT_ENTER_PAGE_INDEX
+ */
+static int
+__page_from_npos_internal(WT_SESSION_IMPL *session, WT_REF **refp, uint32_t flags, double npos)
+{
+    WT_BTREE *btree;
+    WT_DECL_RET;
+    WT_PAGE *page;
+    WT_PAGE_INDEX *pindex;
+    WT_REF *current, *descent;
+    double npos_local;
+    int idx, entries;
+    bool eviction;
+
+    /*
+     * The content of this function is derived from the `random_descent` one but has diverged since.
+     */
+    *refp = NULL;
+
+    btree = S2BT(session);
+    current = NULL;
+    /*
+     * This function is called by eviction to find a page in the cache. That case is indicated by
+     * the WT_READ_CACHE flag. Ordinary lookups in a tree will read pages into cache as needed.
+     */
+    eviction = LF_ISSET(WT_READ_CACHE);
+
+    if (0) {
+restart:; /* Restart the search from the root. */
+    }
+
+    /* Search the internal pages of the tree. */
+    current = &btree->root;
+    npos_local = npos;
+    for (;;) {
+        if (F_ISSET(current, WT_REF_FLAG_LEAF))
+            goto done;
+
+        /* The entire for loop's body is for INTERNAL pages only */
+
+        page = current->page;
+        WT_INTL_INDEX_GET(session, page, pindex);
+        entries = (int)pindex->entries;
+
+        npos_local *= entries;
+        idx = (int)npos_local;
+        idx = WT_CLAMP(idx, 0, entries - 1);
+        npos_local -= idx;
+        descent = pindex->index[idx];
+
+        if (eviction) {
+            /*
+             * In case of eviction, we never want to load pages from disk. Also, page_swap with
+             * WT_READ_CACHE will fail anyway and we'll lose our pointer, so avoid making a call
+             * that will fail.
+             */
+            switch (WT_REF_GET_STATE(descent)) {
+            case WT_REF_DISK:
+            case WT_REF_LOCKED:
+            case WT_REF_DELETED:
+                /* Can't go down from here but it's ok to return this page */
+                goto done;
+            default: /* WT_REF_MEM, WT_REF_SPLIT */
+                goto descend;
+            }
+            /* Unreachable */
```

Comment:
```suggestion
            /* Unreachable */
            __wt_abort(session);
```
Similarly for the other `/* Unreachable */` sections below

Reply:
Hmm it's a great idea to do it if it's important that this point is unreachable. In this particular case, these are just notes for developers reading this code. Nothing bad happens if execution reaches any of these points.
I however understand the danger of such comment in case the code changes in future.
I need to think more how to deal with this.